File: app/src/main/java/com/iffy/mianshi/basic/fanxing/rx/ImplObservable.java
package com.iffy.mianshi.basic.fanxing.rx;

/**
 * author : iffy
 * time   : 2020/03/12
 */
//泛型声明在类上 类里面都可以使用
public class ImplObservable<T> implements Observable<T> {
    T t;

    public ImplObservable(T t) {
        this.t = t;
    }

    public static <T> Observable<T> create(T t) {
        return new ImplObservable<>(t);
    }

    @Override
    public T call() {
        return null;
    }

    @Override
    public <U> Observable<U> map(Func1<T, U> func1) {
        Observable<U> trans = create(func1.trans(t));
        return trans;
    }

    @Override
    public void doOnNext(Action<T> action) {
        action.CallAction(t);
    }


}

File: app/src/main/java/com/iffy/mianshi/pattern/II观察者模式/test/ImplObservable.kt
package com.iffy.mianshi.pattern.II观察者模式.test

/**
 *     author : iffy
 *     time   : 2020/03/12
 */
class ImplObservable:Observable {
    val list:ArrayList<Observer> = ArrayList<Observer>()

    override fun regist(observer: Observer) {
        list.add(observer)
    }


    override fun unregist(observer: Observer) {
        list.remove(observer)
    }

    override fun notifyThem(s: String) {
        list.forEach{
            it.onMessageReceived("$it+ $s")
        }

    }
}